src/wp-includes/block-template-utils.php

```diff
@@ -724,6 +724,83 @@ function _wp_build_title_and_description_for_taxonomy_block_template( $taxonomy,
 	return true;
 }
 
+/**
+ * Build a block template object for based on a WP_Post object.
+ *
+ * @since 6.5.0
+ * @access private
+ *
+ * @param WP_Post $post Template post.
+ * @param array $additional_fields Additional fields to add to the template object.
+ * @param bool $hook_blocks Whether to insert hooked blocks into the content or not.
+ * @return WP_Block_Template|WP_Error Template or error object.
+ */
+function _build_block_template_object_from_wp_post_object( $post, $additional_fields = array() ) {
+	$default_template_types = get_default_block_template_types();
+
+	if ( empty( $additional_fields['theme'] ) ) {
+		return new WP_Error( 'template_missing_theme', __( 'No theme is defined for this template.' ) );
+	}
+	$theme = $additional_fields['theme'];
+
+	$template_file  = _get_block_template_file( $post->post_type, $post->post_name );
+	$has_theme_file = get_stylesheet() === $theme && null !== $template_file;
+	$template_id    = ! empty( $theme ) && ! empty( $post->post_name ) ? $theme . '//' . $post->post_name : null;
```

Comment:
Noting that this [used to use](https://github.com/WordPress/wordpress-develop/pull/6278/files#diff-1d8e05d3ea69557ca8742330e62c1893a72d4ca4ff2ed4ccc5b94bcf5d20fed7L766) `$parent_post->post_name` to build the ID.

This _might_ be significant, as it was originally changed to do so by https://github.com/WordPress/wordpress-develop/pull/3533/files#diff-1d8e05d3ea69557ca8742330e62c1893a72d4ca4ff2ed4ccc5b94bcf5d20fed7R761.

Reply:
I think I may have lost this during a rebase but you're right we should keep this logic in place I feel. My understanding is that when a post revision is saved it fetches the canonical post. Calling it parent post in the code feels a bit confusing too.